File: Fortran/IBM/execute/F2018/cintrop_ts29113/AssumedType/unit_tests/AssumedTypeObj04f.c
/**********************************************************************
*  ===================================================================
*  XL Fortran Test Case                          IBM INTERNAL USE ONLY
*  ===================================================================
*
*  TEST CASE TITLE            : AssumedTypeObj04f
*
*  PROGRAMMER                 : Ren, Jian Gang
*  DATE                       : Apr 14, 2012
*  ORIGIN                     : Linux/AIX Compiler Development,
*                             : IBM Software Solutions China Lab
*
*  PRIMARY FUNCTIONS TESTED   : C-interop Assumed-type object
*                                                   
*  SECONDARY FUNCTIONS TESTED : None 
*
*  DRIVER STANZA              : xlc
*  REQUIRED COMPILER OPTIONS  : 
*
*
*  DESCRIPTION                : Calling a non-BIND(C) procedure from Fortran
*                               where the procedure is defined in C with
*                               assumed-type object. 
*
*1234567890123456789012345678901234567890123456789012345678901234567890*/

#include <stdio.h>

void c_func_add(void* a, void* b)
{
  int v1 = *(int *)a;
  int v2 = *(int *)b; 
  
  printf("(int)a + b: %d\n", v1 + v2);
  
  return;
}

void c_func_arr_sum(void* a, void* b, void* len)
{
  int i = 0;   
  int ln;
  int res = 0;

  int *arr1 = (int *)a; 
  int *arr2 = (int *)b;
  
  ln = *(int *)len;

  for (;i < ln;i++)
  { 
    res += (arr1[i] + arr2[i]);
  }
  printf("Result is : %d\n", res);
  return;
}

File: Fortran/IBM/execute/F2018/cintrop_ts29113/AssumedType/unit_tests/AssumedTypeObj04f.f90
!**********************************************************************
!*  ===================================================================
!*
!*  DATE                       : Apr 14, 2012
!*                             : IBM Software Solutions China Lab
!*
!*  PRIMARY FUNCTIONS TESTED   : C-interop Assumed-type object
!*
!*  SECONDARY FUNCTIONS TESTED : None
!*
!*  REQUIRED COMPILER OPTIONS  :
!*
!*  DESCRIPTION                : Calling a BIND(C) procedure from Fortran
!*                               where the procedure is defined in C with
!*                               assumed-type object.
!*
!234567890123456789012345678901234567890123456789012345678901234567890
program AssumedTypeObj04f
  use, intrinsic :: iso_c_binding
  implicit none

  integer(c_int) :: a = 3
  integer(c_int) :: b = 5
  integer(c_int) :: arr1(3) = 3
  integer(c_int) :: arr2(3) = 4

  interface
    subroutine c_func_add(a, b) BIND(c)
      implicit none

      TYPE(*) :: a, b
    end subroutine c_func_add
    subroutine c_func_arr_sum(a, b, len) BIND(c)
      implicit none

      TYPE(*) :: a(*), b(*), len
    end subroutine c_func_arr_sum
  end interface

  call sub(a, b)
  call arr_add(arr1, arr2, a)

  contains
    subroutine sub(a, b)
      TYPE(*) :: a
      TYPE(*) :: b

      call c_func_add(a, b)
    end

    subroutine arr_add(a, b, len)
      TYPE(*) :: a(*)
      TYPE(*) :: b(*)
      TYPE(*) :: len

      call c_func_arr_sum(a, b, len)
    end
end
